src/common/rsvc.c

```diff
@@ -142,15 +144,21 @@ rsvc_client_process_error(struct rsvc_client *client, int rc,
 			ep->ep_rank, DP_RC(rc));
 	} else if (client->sc_leader_known && client->sc_leader_aliveness > 0 &&
 		   ep->ep_rank == client->sc_ranks->rl_ranks[leader_index]) {
+		/* A leader stepping down replies without hint (so give up).
```

Comment:
I will take another look at this scenario and make sure the logic does what is necessary.

However the logic change here in rsvc_client_process_error() is important for the next case mentioned in the comment: a new leader in the process of stepping up. We need to avoid the "give up the leader" branch here at this moment so that after rsvc_client_process_hint() recognizes "new hint from rank" scenario and sets the aliveness for this new leader to 1, and so that the next client RPC rsvc_client_choose() will try this leader again, rather than cycle to the next in the list.

Reply:
The leader is given up by setting/decrementing sc_leader_aliveness to 0; the "gave up this leader" branch only prepares sc_next. And, rsvc_client_choose only uses sc_next if sc_leader_aliveness is zero. So after your rsvc_client_process_hint resets sc_leader_aliveness back to nonzero, rsvc_client_choose shall not be affected the sc_next change at all, ..., no?

Reply:
See latest patch that modifies the hint processing only. Retested and (unless I'm forgetting a scenario - this logic is complex) it works.